program: Add TestWithOptions() API

danobi opened the pull request:
This PR adds support for a new `TestWithOptions()` API
for `Program`s. The extra options are useful for more advanced
testing.

lmb commented:
Thank you for your PR, I'll try to do a full review tomorrow. In the meantime, please take a look at #576 which discusses adding this feature.

lmb commented on prog_test.go:
```diff
@@ -90,6 +91,56 @@ func TestProgramRun(t *testing.T) {
+		DataEnd: 14,
+	}
+	xdpOut := sys.XdpMd{}
+	ctx := (*(*[unsafe.Sizeof(xdp)]byte)(unsafe.Pointer(&xdp)))[:]
```
See my comment on `binary.Write` above, usage of `unsafe` should not be necessary / encouraged.

dxuuu commented on prog.go:
```diff
@@ -549,6 +549,16 @@ func (p *Program) Close() error {
+	CtxOut []byte
+	Repeat uint32
+	Flags  uint32
+	Cpu    uint32
```
Seems like some prog types (raw tp) setting CPU tells the handler to run the prog on that specific CPU by sending an IPI.

danobi commented:
Pushed a commit that addresses all comments except for RunOptions/RunResult split

danobi commented on prog.go:
```diff
@@ -549,6 +549,16 @@ func (p *Program) Close() error {
+type TestOptions struct {
+	In     []byte
+	Ctx    []byte
+	CtxOut []byte
```
I'm not an expert on go, but how would you convert from `[]byte` -> `interface{}` for returned context?

ti-mo replied:
When working with converting Go types to/from binary, my mind immediately goes to `encoding.Binary(Un)Marshaler`. We could take a `BinaryMarshaler` for `Context`.

The following needs to be taken into account:
```
 *		**BPF_PROG_TYPE_SK_LOOKUP**
 *			*data_in* and *data_out* must be NULL.
 *
 *		**BPF_PROG_TYPE_RAW_TRACEPOINT**,
 *		**BPF_PROG_TYPE_RAW_TRACEPOINT_WRITABLE**
 *
 *			*ctx_out*, *data_in* and *data_out* must be NULL.
 *			*repeat* must be zero.
```

With the current patch, I don't think these prog types will work, since all fields (data_in/out, ctx_in/out) unconditionally receive slice pointers, even if their lengths are zero. This will need to be changed to conditionally set the respective fields when their input slices are > len 0.

On the bright side, this constraint might guide us in designing the API:

```go
type RunContext struct {
	// Program's data input. Required field.
	DataIn []byte

	// Program's context input. Optional field.
	ContextIn encoding.BinaryMarshaler

	// Program's context after Program has run. Optional field.
	ContextOut encoding.BinaryUnmarshaler
	// .. or
	ContextOut []byte
// ...
}
```

- DataOut could be the return value of the function or caller-provided. Both could work given the input struct is called `RunContext`. The forced allocation of the output buffer w/ padding could be moved to `Test()` and we don't need to break API. `Benchmark()` could then run with SkipOutput implicitly enabled.
- If ContextIn is provided, a buffer for ContextOut can be allocated with the same length as the marshaling result. If ContextOut is not provided, this buffer would not be allocated and the CtxOut arg not provided to the syscall to accomodate for `RAW_TRACEPOINT*`.

danobi replied:
With your proposed `RunContext`, we still need to give a way for the user to provide the context value, right? `ContextIn` and `ContextOut` are interfaces. We still need the data somewhere. For that reason, I think it might be cleaner to have `Context` and `ContextOut` to be `interface{}`.

> With the current patch, I don't think these prog types will work, since all fields (data_in/out, ctx_in/out) unconditionally receive slice pointers, even if their lengths are zero.

I think even without this patch, BPF_PROG_TYPE_RAW_TRACEPOINT* was broken: https://github.com/cilium/ebpf/blob/e0825791eaf28e70c8c6403054c85b5d37450290/prog.go#L585-L587 . `data_in` was never permitted to be null.

danobi replied:
I'd like to also better understand the concern about allocating for `Benchmark()`. The prog iterations are run inside the kernel, so allocation time is not measured.

lmb replied:
> When working with converting Go types to/from binary, my mind immediately goes to encoding.Binary(Un)Marshaler.

I'm not a fan of either interface. Because they are usually implemented manually they invite bugs and bloat the code base. BinaryMarshaler also doesn't allow controlling allocation, which really should be possible if the reason to use BinaryMarshaler is performance in the first place.

Hence, I think that interface{} is actually better than the marshalers, since it allows us to cut out all that junk in the common case.

> I think even without this patch, BPF_PROG_TYPE_RAW_TRACEPOINT* was broken.

That's correct. I think Timo is saying that our new API should allow passing NULL in some fashion. We can do that as a follow up PR though, I think.

> I'd like to also better understand the concern about allocating for Benchmark().

The issue is that we allocate an output slice even if the caller discards it: https://github.com/cilium/ebpf/blob/e0825791eaf28e70c8c6403054c85b5d37450290/prog.go#L602

danobi commented:
Not sure what's up with CI. Seems like one of the jobs got stuck (can't get the page to load)

danobi commented:
I've cleaned up the commits and changed `ContextOut` to `interface{}`. IMO this approach balances ergonomics with reflection. Please let me know what you think.

lmb commented on prog.go:
```diff
@@ -506,7 +523,23 @@ func (p *Program) Close() error {
+// Run runs the Program in kernel with given RunOptions.
+//
+// Note: the same restrictions from Test apply.
+func (p *Program) Run(opts RunOptions) (uint32, []byte, error) {
```
See the discussion on `DataOut`, this should not return `[]byte`.

lmb commented on prog.go:
```diff
@@ -498,6 +498,23 @@ func (p *Program) Close() error {
+	// CPU to run Program on. Optional field.
+	// Note not all program types support this field.
+	CPU uint32
+}
```
This should include `Reset func()` with appropriate documentation.

lmb commented on prog_test.go:
```diff
@@ -90,6 +90,54 @@ func TestProgramRun(t *testing.T) {
+	opts := RunOptions{
+		Data:       buf,
+		Context:    xdp,
+		ContextOut: &xdpOut,
```
Could this just be `&xdp` as well?

danobi replied:
It could, but I had it this way to be extra sure that serialization/deserialization path was being hit.

lmb commented on prog_test.go:
```diff
@@ -90,6 +90,54 @@ func TestProgramRun(t *testing.T) {
+	}
+
+	if xdp != xdpOut {
+		t.Errorf("Expect xdp (%+v) != xdpOut (%+v)", xdp, xdpOut)
```
Incorrect error message?

lmb commented on prog_test.go:
```diff
@@ -90,6 +90,54 @@ func TestProgramRun(t *testing.T) {
+		t.Error("Expected return value to be 0, got", ret)
+	}
+
+	if xdp != xdpOut {
```
It might be better to assert that we actually decoded the kernel output ctx. Right now this doesn't really test anything.

danobi replied:
Hmm could you clarify? The current code tests that `DataEnd` was decoded as 14. Cuz when it went into the `Run`, `xdpOut.DataEnd == 0`.

lmb commented on prog.go:
```diff
@@ -498,6 +498,23 @@ func (p *Program) Close() error {
 }
 
+// Various options for Run'ing a Program
+type RunOptions struct {
```
Name bikeshed. Timo suggested `RunContext`, but then we'd have `RunContext.Context`. `RunOptions` is consistent with other option structs we have in the library, but it's a little bit weird in that we actually write into the `Out` fields. Probably not a big deal, so I'm fine with `RunOptions`.

lmb commented on prog.go:
```diff
@@ -577,37 +634,43 @@ var haveProgTestRun = internal.FeatureTest("BPF_PROG_TEST_RUN", "4.12", func() e
+		if err := binary.Write(ctx, internal.NativeEndian, opts.Context); err != nil {
+			return 0, 0, fmt.Errorf("cannot serialize context: %v", err)
+		}
+		if opts.ContextOut != nil {
```
Nit: why make this dependent on `.Context`? This can be a stand alone if branch i think.

danobi replied:
Oh right, cuz I was relying on `ctx.Bytes()` to figure out ctxOut's size. Updated PR with reflection to get size.

lmb commented on prog.go:
```diff
@@ -617,28 +680,37 @@ func (p *Program) testRun(in []byte, repeat int, reset func()) (uint32, []byte,
+			// and the kernel wrote past the end of our buffer.
+			panic("kernel wrote past end of output buffer")
+		}
+		opts.DataOut = opts.DataOut[:int(attr.DataSizeOut)]
```
This has no effect since `opts` is passed by value. `Run`, etc. should take `*RunOptions` I think.

lmb commented on prog.go:
```diff
@@ -498,6 +498,29 @@ func (p *Program) Close() error {
+	// CPU to run Program on. Optional field.
+	// Note not all program types support this field.
+	CPU uint32
+	// Called whenever the syscall is interrupted, and should be set to testing.B.ResetTime
```
```suggestion
	// Called whenever the syscall is interrupted, and should be set to testing.B.ResetTimer
```

lmb commented on prog.go:
```diff
@@ -498,6 +498,29 @@ func (p *Program) Close() error {
+	// Program's data input. Required field.
+	Data []byte
+	// Program's data after Program has run. Caller must allocate. Optional field.
+	// If the data output from the kernel is larger than the provided buffer, Run will panic.
```
I think we should keep this as an implementation detail and not document it. On newer kernels we will get ENOSPC, and on older kernels only programs that extend the buffer may trigger this.

lmb reviewed (changes requested):
Looks great, thanks!

lmb commented:
Hi @danobi, I'll only be able to take another look at this towards Wednesday afternoon, European timezone. Sorry!

lmb commented on prog.go:
```diff
@@ -565,37 +622,44 @@ var haveProgTestRun = internal.FeatureTest("BPF_PROG_TEST_RUN", "4.12", func() e
-	out := make([]byte, len(in)+outputPad)
+	var ctxOut []byte
+	if opts.ContextOut != nil {
+		ctxOut = make([]byte, reflect.Indirect(reflect.ValueOf(opts.ContextOut)).Type().Size())
```
Using reflection here is incorrect, since the size of a type in memory != the size of its encoded form. For example, go type may contain padding due to alignment which binary.Size will ignore.

```suggestion
		ctxOut = make([]byte, binary.Size(opts.ContextOut))
```

lmb commented on prog.go:
```diff
@@ -565,37 +622,44 @@ var haveProgTestRun = internal.FeatureTest("BPF_PROG_TEST_RUN", "4.12", func() e
 
-	if err := haveProgTestRun(); err != nil {
-		return 0, nil, 0, err
+	ctx := new(bytes.Buffer)
```
Nit: this forces an allocation even if we don't have opts.Context. We can fix that in a follow up though.

dxuuu replied:
Ok, will send followup

lmb reviewed (changes requested):
This looks really nice, thanks for putting in so much work. I've found one problem, once that is gone this is good to go.

ti-mo commented on prog.go:
```diff
@@ -565,37 +622,44 @@ var haveProgTestRun = internal.FeatureTest("BPF_PROG_TEST_RUN", "4.12", func() e
-	if uint(repeat) > math.MaxUint32 {
-		return 0, nil, 0, fmt.Errorf("repeat is too high")
+func (p *Program) testRun(opts *RunOptions) (uint32, time.Duration, error) {
+	if len(opts.Data) == 0 {
```
data_in must be zero for `SK_LOOKUP`, `RAW_TRACEPOINT` and `RAW_TRACEPOINT_WRITABLE`, so let's not make the assumption that `Data` is/needs to be populated.

dxuuu replied:
Yeah I agree. I think there was some discussion above. This was a carry-over from existing behavior, so probably best to change in a separate PR. I can file a followup.

ti-mo reviewed (commented):
A few remaining comments, probably ok to merge after these.

lmb commented:
Thanks for your contribution, sorry it took so long to get merged!
